airbyte-integrations/connectors/source-myhours/source_myhours/source.py

```diff
@@ -0,0 +1,172 @@
+    ) -> str:
+        return "Projects/getAll"
+
+
+class Tags(MyhoursStream):
+    primary_key = "id"
+
+    def path(
+        self, stream_state: Mapping[str, Any] = None, stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> str:
+        return "Tags"
+
+
+class TimeLogs(MyhoursStream):
+    primary_key = "logId"
+
+    def __init__(
+        self,
+        authenticator: TokenAuthenticator,
+        start_date: str,
+        batch_size: int,
+        **kwargs,
+    ):
+        self.start_date = pendulum.parse(start_date)
+        self.batch_size = batch_size
+
+        if self.start_date > pendulum.now():
+            self.logger.log(logging.WARN, f'Stream {self.name}: start_date "{start_date.isoformat()}" should be before today.')
+
+        super().__init__(authenticator=authenticator)
+
+    def path(
+        self, stream_state: Mapping[str, Any] = None, stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> str:
+        return "Reports/activity"
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        previous_query = parse_qs(urlparse(response.request.url).query)
+        previous_end = pendulum.parse(previous_query["DateTo"][0])
+
+        new_from = previous_end.add(days=1)
+        new_to = new_from.add(days=self.batch_size - 1)
+
+        if new_from > pendulum.now():
+            return None
+
+        return {
+            "DateFrom": new_from.to_date_string(),
+            "DateTo": new_to.to_date_string(),
+        }
+
+    def request_params(
+        self,
+        stream_state: Mapping[str, Any],
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> MutableMapping[str, Any]:
+
+        if next_page_token is None:
+            return {"DateFrom": self.start_date.to_date_string(), "DateTo": self.start_date.add(days=6).to_date_string()}
+        return next_page_token
+
+
+class Users(MyhoursStream):
+    primary_key = "id"
+
+    def path(
+        self, stream_state: Mapping[str, Any] = None, stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> str:
+        return "Users/getAll"
+
+
+# Source
+class SourceMyhours(AbstractSource):
+    def get_access_token(self, config) -> Tuple[str, any]:
+        email = config["email"]
+        password = config["password"]
+        url = f"{url_base}/tokens/login"
+
+        try:
+            payload = json.dumps({"grantType": "password", "email": email, "password": password, "clientId": "api"})
+            response = requests.post(url, headers=request_headers, data=payload)
+            response.raise_for_status()
+            json_response = response.json()
+            return json_response.get("accessToken", None), None if json_response is not None else None, None
```

Comment:
Since this is my first time using Python, I copied this from another connector and couldn't wrap my head around it. I thought this line of python code was too advanced for me to comprehend 😂 Glad it can be removed.